### ISSUE TITLE
Reproduce Push-T results reported in the paper

### ISSUE THREAD

**Issue by fedeceola** [HRI-EU/flow_matching#3]
Hi, thanks for your work!

I am trying to reproduce the results in your paper with the code in this repository. I downloaded the pre-trained Push-T weights and computed the average final rewards as explained in the paper, obtaining a reward ~0.56, while in the paper you report a reward equal to 0.6218. I also retrained the model a couple of times, obtaining a reward in both cases ~0.51.

I had to modify this line

https://github.com/HRI-EU/flow_matching/blob/6ba6b81f08d80a37157b03b86400c2b879e2043a/examples/flow_pusht.py#L210

to set the value to 16. And add a couple of lines of code to compute the final rewards.

Am I missing anything?

**Comment by fedeceola** [comment]
Hi @fan6zh, sure! Here you can find my modified test function from flow_pusht.py:

```python
def test():

    PATH = './checkpoint_t/flow_ema_03000.pth'
    state_dict = torch.load(PATH, map_location='cuda')
    ema_nets = nets
    ema_nets.vision_encoder.load_state_dict(state_dict['vision_encoder'])
    ema_nets.noise_pred_net.load_state_dict(state_dict['noise_pred_net'])

    max_steps = 300
    env = pusht.PushTImageEnv()

    test_start_seed = 1000
    n_test = 500

    rewards_pp_avg = []
    ###### please choose the seed you want to test
    for epoch in range(n_test):
        seed = test_start_seed + epoch
        env.seed(seed)
        final_rewards = []

        for pp in range(10):
            obs, info = env.reset()
            obs_deque = collections.deque(
                [obs] * obs_horizon, maxlen=obs_horizon)
            imgs = [env.render(mode='rgb_array')]
            rewards = list()
            done = False
            step_idx = 0

            with tqdm(total=max_steps, desc="Eval PushTImageEnv") as pbar:
                while not done:
                    B = 1
                    x_img = np.stack([x['image'] for x in obs_deque])
                    x_pos = np.stack([x['agent_pos'] for x in obs_deque])
                    x_pos = pusht.normalize_data(x_pos, stats=stats['agent_pos'])

                    x_img = torch.from_numpy(x_img).to(device, dtype=torch.float32)
                    x_pos = torch.from_numpy(x_pos).to(device, dtype=torch.float32)
                    # infer action
                    with torch.no_grad():
                        # get image features
                        # t1 = time.time()
                        image_features = ema_nets['vision_encoder'](x_img)
                        obs_features = torch.cat([image_features, x_pos], dim=-1)
                        obs_cond = obs_features.unsqueeze(0).flatten(start_dim=1)

                        timehorion = 16
                        for i in range(timehorion):
                            noise = torch.rand(1, pred_horizon, action_dim).to(device)
                            x0 = noise.expand(x_img.shape[0], -1, -1)
                            timestep = torch.tensor([i / timehorion]).to(device)

                            if i == 0:
                                vt = nets['noise_pred_net'](x0, timestep, global_cond=obs_cond)
                                traj = (vt * 1 / timehorion + x0)

                            else:
                                vt = nets['noise_pred_net'](traj, timestep, global_cond=obs_cond)
                                traj = (vt * 1 / timehorion + traj)

                    # print(time.time() - t1)

                    naction = traj.detach().to('cpu').numpy()
                    naction = naction[0]
                    action_pred = pusht.unnormalize_data(naction, stats=stats['action'])

                    # only take action_horizon number of actions
                    start = obs_horizon - 1
                    end = start + action_horizon
                    action = action_pred[start:end, :]

                    # x_img = x_img[0, :].permute((1, 2, 0))
                    # plot_trajectory(x0[0].detach().cpu().numpy(), vt[0].detach().cpu().numpy(),
                    #                 action_pred,
                    #                 x_img.detach().cpu().numpy())

                    # execute action_horizon number of steps
                    for j in range(len(action)):
                        # stepping env
                        obs, reward, done, _, info = env.step(action[j])
                        # save observations
                        obs_deque.append(obs)
                        # and reward/vis
                        rewards.append(reward)
                        imgs.append(env.render(mode='rgb_array'))

                        # update progress bar
                        step_idx += 1

                        pbar.update(1)
                        pbar.set_postfix(reward=reward)

                        if step_idx > max_steps:
                            done = True
                        if done:
                            break
            final_rewards.append(rewards[-1])
        rewards_pp_avg.append(sum(final_rewards)/len(final_rewards))
        print("Average pp rewards:", rewards_pp_avg[epoch])
    print("Average reward", sum(rewards_pp_avg)/len(rewards_pp_avg))
```

**Comment by fan6zh** [comment]
Hi, I am using the maximum target coverage, as in the diffusion policy colab. I think that causes the difference.
